### DISCUSSION TITLE
mosfets

### DISCUSSION THREAD

**Discussion by POP1PO** [florindumitrescu94/DashBoard_PowerBox_V2#4] [Ideas]
Could the control of some MOSFETs be added to turn devices on or off at will?

**Comment by florindumitrescu94** [comment]
You mean like controlling each DC output separately? 
It can be designed to do this, there are plenty of pins left on the arduino, but it would require changing everything, the driver, arduino code and schematic. That is beyond the scope of this project (for me, at least).
However, feel free to use this project as a starting point and design your own circuit and code that would fit your needs.
Cheers!

**Reply by gvalsvik** [reply]
Hello @florindumitrescu94,

I'm also quite interested in having individual control of each of the 4 power outlets. 
I have done quite a bit of Arduino programming, but don't know anything on how to change the Ascom driver.
Could you point me to what software I need to use and what project file to change that has the ASCOM driver?

Thanks.

**Comment by POP1PO** [comment]
Do you have the schematics available for any type of design software?

**Reply by florindumitrescu94** [reply]
Certainly! Here's the project file for EasyEDA. Install EasyEDA, download this file: https://we.tl/t-4BuR5aDgch and in EasyEDA click File>Open>EasyEDA and select this json file.

**Comment by POP1PO** [comment]
I have modified the scheme to have control over the 4 independent outputs.
I have also removed all the remaining pins from the nano to use them in the future.

I think everything is fine except for some kind soul to confirm it in case I missed something.




![Schematic_DashBoard-powerbox_2024-11-03](https://github.com/user-attachments/assets/48d5a29e-919d-4e3f-a026-216c746c9f1a)

**Reply by florindumitrescu94** [reply]
Hello,

Your schematic would not function correctly. You have skipped adding the driver NPN transistor for each MOSFET. The voltage at the gate needs to be the same as the one on the source. 
When you're switching the Arduino pin to Low, it shunts to ground, which would turn the P-MOSFET on, but when you turn it to High, it provides only 5V, so the MOSFET will stay on, never turning off. 

Furthermore, the signal you're providing from the Arduino is not even going to the gate of the MOSFET... it looks like you've connected it to the source. Maybe I am not understanding correctly, but what exactly are you trying to acheive by doing this?
I've tried running a simulation of this circuit and it does not work. 

I know adding four extra transistors in the mix might seem like a chore, but it's a necessary one. 

The way it works is as follows: 
Case one: 
On the left, we have what's inside a microcontroller: A pull-up resistor (represented by the 40k resistor in this case) and a switch to ground (basically that how LOW and HIGH is created inside a MCU (microcontroller). In this first case, the pin is HIGH, providing 5V to the NPN transistor base. This, in turn, opens the NPN transistor, which now acts as a direct path to ground for the 12 volts coming throught the 10k resistor (the one connected between the source and the gate of the P-MOSFET.
That means that, now, the voltage at the P-MOSFET gate is close to 0V. This also means that the voltage difference between the gate and the source is almost the same as the voltage at the source. This opens up the MOSFET, allowing current to flow between its source and drain through our load, represented by the 1k resistor between drain and ground.


![image](https://github.com/user-attachments/assets/39c5dcbf-34ee-4543-a6a7-26ee7929dc8c)

Case 2: 
Same circuit, but the pin is now LOW.
The voltage supplied to the base of the NPN transistor is now 0, so it is closed. 
This, in turn, makes the voltage at the gate of the P-MOSFET to be almost the same as the voltage at the source (voltage difference between gate and source is close to 0). Therefore, the MOSFET does not allow any current flow. 

![image](https://github.com/user-attachments/assets/cfe1daa1-f15f-4011-9097-5cf76556b9de)



BTW, an explanation for why I have not used an N-MOSFET: The N-MOSFET needs to be between the load and ground, not between the V+ and load, as it is the case with the P-MOSFET. That causes some issues, because some devices (like a cooled camera) is connected to ground already via the USB cable, so things don't shut down at all, because the current can always find a path to ground, rendering the switch on the negative side useless. 
N-MOSFETs work fine for PWM control, because we don't have that alternative path to ground on the dew heaters. 

Adding that header there sure is a good idea, since it enables you to add new features in the future. I think you could pretty easily add a circuit to control a focuser or PWM control for a Flats panel, or switch for a dome, or automated flats panel, relay for a 230/120V system... the ideas are endless, provided you write them in the code, as well.

Also, love your idea to use GX connectors instead of DC jacks... I wish I could do that, too, but I wanted to keep it as compact as possible, since mine is forever connected on the telescope and never comes down. I did employ GX20 connectors on my main battery, tho. I might revisit this power box in the future, as my needs change and, at that time, I will go with GX12 connectors, for sure.

I hope this helps!  :)

**Reply by gvalsvik** [reply]
Hello,

Great suggestion @POP1PO.  Did you do anything on the implementation of the 4port individual switching option? I mean like updated Arduino code and ASCOM driver. 
I'm looking for the exact same thing.  I probably could do the arduino code changes, but I have no clue on how to write the ASCOM driver.  Thanks.

**Reply by florindumitrescu94** [reply]
I'm currently working on the V3 version of this power box, which will have 4 individually controlled ports (Ports 1 2 and 3 will be individually controlled, while 4 and 5 will be linked together), an Always On port and two external controls through optocouplers. I'm currently waiting for the parts to arrive so I can test it, but the software part is mostly ready. I've also built a control app that lets you use the switches independent of any astro software like NINA or APT. Stay tuned!

**Reply by gvalsvik** [reply]
That's great news @florindumitrescu94 !  Very much looking forward to the V3, looks like it is exactly what I need.
I see there will be an always on port. This can be used to power my Mele control PC :)

Thank!

**Reply by florindumitrescu94** [reply]
Glad you're excited for it! Yes, that's exactly what I'm going to use it on my setup. All ports pass through the current sensor, so the power usage will be recorded, too. Have you been using the Auto PWM function? I've elected to remove the two NTC probes from the V3 because I found that the functionality could be erratic. I've, instead, made a function that sets the PWM power based on the difference between ambient and calculated dew point, so that anyone can use the Auto PWM without having to use the NTC probes.

**Reply by gvalsvik** [reply]
Hi @florindumitrescu94,
So far I have not used the Auto PWM function. I normally just set the PWM % to a fairly low value and keep them there.  Of course I see the point when using the rig on battery power, but so far I have only used it where mains power is available.

One thing I would like to comment if you have not already changed this:
(I'm using the PowerBox v2 in Nina, so my comments are based on that.)

When powering up the PowerBox, the Dew heater duty cycle will always start at 0%. I would really like to have the previous setting saved, so that the previous duty cycle % is loaded.  

Also, I see that in Nina, when disabling the Switch device, or closing Nina altogether, the Dew Heaters will remain on. Should do a small change in ASCOM driver, so that when the Switch device is disconnected in Nina, it will also set the Dew Heater duty cycle to 0.
When connecting the Switch again, as it is currently in V2, the Dew Heaters will be set to 0%. 

I think it would be more logical to do the opposite:
When Switch device is disconnected: Set Duty cycle to 0%
When Switch device is connected: Set Duty cycle to the previously used %

Thanks and regards :)

**Reply by florindumitrescu94** [reply]
Hi

Since I've built my V2, I've made some mods to it not captured in the project description (since I had in mind to build a V3 from the start, so I've used the V2 as a testing platform). I've added a cap across RST and GND to keep the Arduino from resetting when disconnected from NINA. I guess I could add a function to write the previous value to the EEPROM, so that, when powering the box off and back on, it could read the values from the EEPROM and set them accordingly. However, the problem with that would be that, whenever the device disconnects from NINA, the power gets set to 0, so that is actually the most recently used value... I'll explore this option and see what I can come up with.

**Reply by gvalsvik** [reply]
Hi @florindumitrescu94,

My experience with the dew heaters is that they do not go to 0 power when the device disconnects in Nina. The power is still going, so if you write these values to EPROM before it disconnects, they will be available when the device reconnects.

Thanks.

**Reply by florindumitrescu94** [reply]
Thanks for the suggestion! 
I'll see what I can do once I have built the board and can test the driver. I'm currently waiting for parts to arrive from aliexpress. 
I've also included the values being saved in the Driver Profile in Registries and I'll have to see how that works. 

![image](https://github.com/user-attachments/assets/4c7a86b6-c94e-412b-8931-70181e8cf929)

**Reply by gvalsvik** [reply]
Hello @florindumitrescu94,

I was just reading about developing ASCOM drivers and handling persistent data.  You have probably read this yourself, but in case not I copying the info below here. Perhaps this method could be used as well to store the "previous values" for dew heater duty cycle.

Thanks :)



Storage of Driver-Specific Data
Most drivers will need to save persistent data (e.g. COM port selection). It's highly recommended that you use the DriverHelper.Profile object for this. It has a very simple interface and guarantees that you will get private storage for your persistent data items. It uses an area of the registry set aside for ASCOM drivers. A browser, the Profile Explorer, is provided in the Platform to help diagnose ASCOM registration issues and to allow inspection and editing of driver-specific data.

It is possible that the layout of data in the registry will change due to future opperating system restrictions. This already happened twice(!) for Vista and Windows 7. Either way, if you write directly to the registry you're asking for trouble. Drivers which used the Profile object were unaffected by those changes to the Platform. On the other hand, if you choose to write your own private persistence storage code (e.g. to an XML file) that's fine too. Just be aware that your persistent data will not show (and be editable) in the ASCOM Profile Explorer, which is a useful troubleshooting tool.

**Reply by florindumitrescu94** [reply]
Hello,

Thanks for the suggestion. This is exactly what I am going to use and it's what I've captured in the ss above. These are saved in Windows Registries and can be seend using regedit. 

Cheers!

**Reply by florindumitrescu94** [reply]
PowerBox V3 is out now!  Find it here: https://github.com/florindumitrescu94/DashBoard_PowerBox_V3

**Reply by lucvb** [reply]
Hello @florindumitrescu94,
This looks like a fun an interesting project. Thanks for your efforts.
Being not so keen on etching PCB, I would like to contribute and make gerber files. Can you share the schematics?
Regards
